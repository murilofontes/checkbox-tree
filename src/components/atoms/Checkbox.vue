<template>
  <div class="checkbox-container custom-control custom-checkbox">
    <input
      type="checkbox"
      class="custom-control-input"
      :value="value"
      v-model="selected"
      :id="'chkbox-' + value"
      @input="selectAllChildren"
    />
    <label class="custom-control-label" :for="'chkbox-' + value">{{ label }}</label>
  </div>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
export default {
  name: "Checkbox",
  props: {
    label: {
      type: String,
      default: ""
    },
    value: {
      type: String,
      default: ""
    },
    children: {
      type: Object
    }
  },
  data: () => {
    return {
      descendantsIds: []
    };
  },
  methods: {
    ...mapActions(["setName", "setChildrenNames", "unsetChildrenNames"]),
    async selectAllChildren() {
      if (Object.values(this.children).length > 0) {
        await this.getAllDescendantsValues(this.children);
        if (!this.selectedNames.includes(this.value)) {
          this.setChildrenNames(this.descendantsIds);
        } else {
          this.unsetChildrenNames(this.descendantsIds);
          this.descendantsIds = []
        }
      }
    },
    getAllDescendantsValues(children) {
      let _this = this;
      for (const child of Object.values(children)) {
        _this.descendantsIds.push(child.id);
        if (Object.values(child.children).length > 0) {
          _this.getAllDescendantsValues(child.children);
        }
      }
    }
  },
  computed: {
    ...mapGetters(["selectedNames"]),
    selected: {
      get() {
        return this.selectedNames;
      },
      set(value) {
        this.setName(value);
      }
    }
  }
};
</script>

<style lang="scss" scoped>
.checkbox-container {
  cursor: pointer;
  label {
    width: 100%;
    cursor: pointer;
  }
}
</style>