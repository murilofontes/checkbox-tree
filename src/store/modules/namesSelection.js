import
  {
    SET_SELECTED_NAME,
    SET_CHILDREN_NAMES,
    UNSET_CHILDREN_NAMES,
    CLEAR_ALL_SELECTED_NAMES
  } from '../types'

const state = {
  namesSelected: []
}
const mutations = {
  [SET_SELECTED_NAME] (state, id)
  {
    state.namesSelected = id
  },
  [SET_CHILDREN_NAMES] (state, childrenIds)
  {
    state.namesSelected = state.namesSelected.concat(childrenIds)
  },
  [UNSET_CHILDREN_NAMES] (state, childrenIds)
  {
    state.namesSelected = state.namesSelected.filter((id) => !childrenIds.includes(id));
  },
  [CLEAR_ALL_SELECTED_NAMES] (state) {
    state.namesSelected = []
  }
}
const getters = {
  selectedNames: state =>
  {
    return state.namesSelected
  }
}

const actions = {
  setName ({ commit }, id)
  {
    commit(SET_SELECTED_NAME, id)
  },
  setChildrenNames ({ commit }, ids)
  {
    commit(SET_CHILDREN_NAMES, ids)
  },
  unsetChildrenNames ({ commit }, ids)
  {
    commit(UNSET_CHILDREN_NAMES, ids)
  },
  clearAllNamesSelection ({ commit }) {
    commit(CLEAR_ALL_SELECTED_NAMES)
  }
}

export default {
  state,
  mutations,
  actions,
  getters
}
export { mutations, actions, getters }
