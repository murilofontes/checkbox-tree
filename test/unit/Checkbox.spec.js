import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Checkbox from '@/components/atoms/Checkbox'
import { getters } from '@/store/modules/namesSelection'

const localVue = createLocalVue()

localVue.use(Vuex)

let actions = {
  setChildrenNames: jest.fn(),
  unsetChildrenNames: jest.fn(),
  setName: jest.fn()
}

let store = new Vuex.Store({
  actions,
  getters
})

const wrapper = shallowMount(Checkbox, {
  localVue,
  store,
  propsData: {
    label: 'Name',
    value: '123',
    children: {
      "id": "a853dddc-b639-41e6-a876-958b1e7f65d1",
      "name": "Harald Svante August",
      "children": {}
    }
  }
})

describe('Checkbox', () =>
{
  it('Check if component renders corretly"', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('Check if getAllDescendantsValues method is called once its has a children element"', () => {
    jest.spyOn(wrapper.vm, 'getAllDescendantsValues')

    wrapper.find('#chkbox-123').trigger('click')
    expect(wrapper.vm.getAllDescendantsValues).toHaveBeenCalled()
  })
})
