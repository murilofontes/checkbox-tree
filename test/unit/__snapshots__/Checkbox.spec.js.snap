// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Check if component renders corretly" 1`] = `<div class="checkbox-container custom-control custom-checkbox"><input type="checkbox" id="chkbox-123" class="custom-control-input" value="123"> <label for="chkbox-123" class="custom-control-label">Name</label></div>`;
